{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "02ea983c-fb40-42c9-9898-a4a493b8a3b7",
   "metadata": {},
   "source": [
    "# SNAKEMAKE"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 173,
   "id": "69aeb845-dd1a-479b-b5d9-d16efbf1a431",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Overwriting snakefile2\n"
     ]
    }
   ],
   "source": [
    "%%writefile snakefile2\n",
    "from pathlib import Path\n",
    "\n",
    "## This is the new improved pipe for smallseq analysis and filtering. \n",
    "\n",
    "#TODO: CHANGE TO ADD CONFIG FILE AS INPUT! \n",
    "\n",
    "# CHANGE ME\n",
    "READS = [file.stem for file in Path('SRR_new/').iterdir() if file.suffix == '.fastq']\n",
    "\n",
    "rule all:\n",
    "    input:\n",
    "        expand('snakemake_counts/idxstats/mature/{file}_counts_mature.tsv', file = READS),\n",
    "        expand('snakemake_bowtie2/mature/{file}_mature_bowtie2_aligned.bam', file = READS),\n",
    "        expand('snakemake_salmon/mature/{file}_salmon_mature/', file = READS),\n",
    "        expand('snakemake_salmon/transcriptome/{file}_salmon_transcriptome/', file = READS),\n",
    "        expand('snakemake_bowtie2/mature_T/{file}_mature_bowtie2_aligned.bam', file = READS),\n",
    "        'snakemake_counts/featurecounts/genome/featurecounts_genome_clean.txt'\n",
    "        \n",
    "        \n",
    "### MIRBASE ALIGNMENT ###:         \n",
    "        \n",
    "rule remove_umi:\n",
    "    input:\n",
    "        raw_reads = 'SRR_new/{file}.fastq'\n",
    "    output:\n",
    "        umi_reads = 'snakemake_umi/umi_trimmed/{file}_umi.fastq',\n",
    "        log_file = 'snakemake_umi/reports/{file}_umi.log'\n",
    "    shell:\n",
    "        \"umi_tools extract --extract-method=regex --bc-pattern='(?P<discard_1>.*)(?P<umi_1>[ACT]{{8}}CA)' \" \n",
    "        \"-I {input.raw_reads} -S {output.umi_reads} -L {output.log_file}\" \n",
    "        #FOR REMOVING 3' ADAPTER: '(?P<discard_1>.*)(?P<umi_1>[ACT]{{8}}CA).*(?P<discard_2>TGGAATTCT.+)' \n",
    "        \n",
    "# REMOVED -u here because it filters the 5' ends of the read. I have already done this in umi_tools! \n",
    "rule cutadapt:\n",
    "    input:\n",
    "        'snakemake_umi/umi_trimmed/{file}_umi.fastq'\n",
    "    output:\n",
    "        cutadapt_reads = 'snakemake_cutadapt/{file}_umi_cutadapt.fastq',\n",
    "        log = 'snakemake_cutadapt/reports/{file}_cutadapt.log'\n",
    "    params:\n",
    "        adapter_file = 'adapters/adapters.txt'\n",
    "    shell:\n",
    "        \"cutadapt -a file:{params.adapter_file} -e 0.1 -O 1 --quiet --minimum-length 18 \" \n",
    "        \"--maximum-length 40 -o {output.cutadapt_reads} {input} > {output.log}\"\n",
    "\n",
    "# maybe change the --norc to see if the alignment rate goes up? \n",
    "rule bowtie1_mature_alignment:\n",
    "    input:\n",
    "        reads = 'snakemake_cutadapt/{file}_umi_cutadapt.fastq'\n",
    "    output:\n",
    "        aligned_sam = 'snakemake_bowtie/mature/aligned/{file}_bowtie_aligned.sam',\n",
    "        non_aligned = 'snakemake_bowtie/mature/not_aligned/{file}_bowtie_non_aligned.fastq',\n",
    "        log = 'snakemake_bowtie/mature/reports/{file}_bowtie_aligned.log'\n",
    "    params:\n",
    "        index = 'indexes/bowtie_index/hsa_mature_index/hsa_mature'\n",
    "    shell:\n",
    "        \"bowtie -n 0 -l 32 --norc --best --strata -m 1 -x {params.index} {input.reads} \"\n",
    "        \"--un {output.non_aligned} -S {output.aligned_sam} 2> {output.log}\" \n",
    "        \n",
    "rule mature_sam_to_bam:\n",
    "    input:\n",
    "        aligned_sam = 'snakemake_bowtie/mature/aligned/{file}_bowtie_aligned.sam'\n",
    "    output:\n",
    "        sorted_bam = 'snakemake_bowtie/mature/aligned/{file}_bowtie_aligned.sorted.bam'\n",
    "    shell:\n",
    "        \"samtools sort -o {output.sorted_bam} {input.aligned_sam} && rm {input.aligned_sam} \"\n",
    "        \"&& samtools index {output.sorted_bam}\"\n",
    "            \n",
    "rule mirbase_hairpin_alignment:\n",
    "    input:\n",
    "        reads = 'snakemake_cutadapt/{file}_umi_cutadapt.fastq'\n",
    "    output:\n",
    "        aligned_sam = 'snakemake_bowtie/hairpin/aligned/{file}_bowtie_aligned.sam',\n",
    "        non_aligned = 'snakemake_bowtie/hairpin/not_aligned/{file}_bowtie_non_aligned.fastq',\n",
    "        log = 'snakemake_bowtie/hairpin/reports/{file}_bowtie_aligned.log'\n",
    "    params:\n",
    "        index = 'indexes/bowtie_index/hsa_hairpin_index/hsa_hairpin'\n",
    "    shell:\n",
    "        \"bowtie -n 0 -l 32 --norc --best --strata -m 1 -x {params.index} {input.reads} \"\n",
    "        \"--un {output.non_aligned} -S {output.aligned_sam} 2> {output.log}\" \n",
    "\n",
    "rule hairpin_sam_to_bam:\n",
    "    input:\n",
    "        aligned_sam = 'snakemake_bowtie/hairpin/aligned/{file}_bowtie_aligned.sam'\n",
    "    output:\n",
    "        sorted_bam = 'snakemake_bowtie/hairpin/aligned/{file}_bowtie_aligned.sorted.bam'\n",
    "    shell:\n",
    "        \"samtools sort -o {output.sorted_bam} {input.aligned_sam} && rm {input.aligned_sam} \"\n",
    "        \"&& samtools index {output.sorted_bam}\"\n",
    "\n",
    "# Just testing STAR. Really low alignment rate, doesnt align at all.\n",
    "rule star_all_reads:\n",
    "    input:\n",
    "        reads = 'snakemake_cutadapt/{file}_umi_cutadapt.fastq'\n",
    "    output:\n",
    "        'snakemake_star/{file}_Aligned.out.sam',\n",
    "        'snakemake_star/{file}_Log.final.out',\n",
    "        'snakemake_star/{file}_Log.out',\n",
    "        'snakemake_star/{file}_Log.progress.out',\n",
    "        'snakemake_star/{file}_SJ.out.tab'\n",
    "    params:\n",
    "        index = 'indexes/star_index/',\n",
    "        prefix = 'snakemake_star/{file}_'\n",
    "    shell:\n",
    "        \"STAR --runThreadN 4 --genomeDir {params.index} --readFilesIn {input.reads} \" \n",
    "        \"--readFilesCommand zcat --outSAMstrandField intronMotif \"\n",
    "        \"--outFilterMultimapNmax 50 --outFilterScoreMinOverLread 0 --outFilterMatchNmin 18 \" \n",
    "        \"--outFilterMatchNminOverLread 0 --outFilterMismatchNoverLmax 0.04 --alignIntronMax 1 --outFileNamePrefix {params.prefix}\"\n",
    "\n",
    "rule star_sam_to_bam:\n",
    "    input:\n",
    "        aligned_sam = 'snakemake_star/{file}_Aligned.out.sam'\n",
    "    output:\n",
    "        sorted_bam = 'snakemake_star/{file}_aligned.sorted.bam'\n",
    "    shell:\n",
    "        \"samtools sort -o {output.sorted_bam} {input.aligned_sam} && rm {input.aligned_sam} \"\n",
    "        \"&& samtools index {output.sorted_bam}\"\n",
    "        \n",
    "rule deduplicate_umi_mature:\n",
    "    input:\n",
    "        'snakemake_bowtie/mature/aligned/{file}_bowtie_aligned.sorted.bam' \n",
    "    output:\n",
    "        deduplicated = 'snakemake_deduplicated/mature/reads/{file}_deduplicated_mature.bam',\n",
    "        log = 'snakemake_deduplicated/mature/reports/{file}_deduplicated_mature.log'\n",
    "    shell:\n",
    "        \"umi_tools dedup --method=unique -I {input} -S {output.deduplicated} -L {output.log} && \"\n",
    "        \"samtools index {output.deduplicated}\"\n",
    "        \n",
    "# count reads with samtools idxstats from the deduplicated files from the alignment to mirbase mature      \n",
    "rule samtools_count_idxstats_mirbase_mature:\n",
    "    input:\n",
    "        deduplicated = 'snakemake_deduplicated/mature/reads/{file}_deduplicated_mature.bam'\n",
    "    output:\n",
    "        idxstats_count = 'snakemake_counts/idxstats/mature/{file}_counts_mature.tsv'\n",
    "    shell:\n",
    "        \"samtools idxstats {input.deduplicated} | cut -f 1,3 > {output.idxstats_count}\" \n",
    "    \n",
    "    \n",
    "### Align unalign reads that did not align to mirBase to genome ###\n",
    "\n",
    "\n",
    "#Align better than star, around 30-40% alignment. Looser paramters than bowtie1 and to the mirbase mature. \n",
    "rule bowtie2_genome_alignment:\n",
    "    input:\n",
    "        non_aligned = 'snakemake_bowtie/mature/not_aligned/{file}_bowtie_non_aligned.fastq'\n",
    "    output:\n",
    "        aligned = 'snakemake_bowtie2/mirbase_unaligned/{file}_aligned_bowtie2.bam',\n",
    "        log = 'snakemake_bowtie2/mirbase_unaligned/reports/{file}.log'\n",
    "    params:\n",
    "        index = 'indexes/bowtie2_index/genome/GRCh38'\n",
    "    shell:\n",
    "        \"bowtie2 -k 100 --local --very-sensitive-local -x {params.index} -U {input.non_aligned} 2> {output.log} | \" \n",
    "        \"samtools view -b - | samtools sort -o {output.aligned} - && samtools index {output.aligned}\"\n",
    "        \n",
    "# deduplicate umi from the reads aligned to genome with bowtie2     \n",
    "rule deduplicate_umi_bowtie2_alignment:\n",
    "    input:\n",
    "        aligned = 'snakemake_bowtie2/mirbase_unaligned/{file}_aligned_bowtie2.bam' \n",
    "    output:\n",
    "        deduplicated = 'snakemake_deduplicated/genome/reads/{file}_deduplicated_genome.bam',\n",
    "        log = 'snakemake_deduplicated/genome/reports/{file}_deduplicated_genome.log'\n",
    "    shell:\n",
    "        \"umi_tools dedup --method=unique -I {input} -S {output.deduplicated} -L {output.log} && \"\n",
    "        \"samtools index {output.deduplicated}\"\n",
    "\n",
    "# count the deduplicated reads aligned to genome with featureCounts \n",
    "rule count_reads_genome_featurecounts:\n",
    "    input:\n",
    "        gtf_file = 'genomes/gencode.v38.chr_patch_hapl_scaff.annotation.gtf'\n",
    "    params:\n",
    "        reads_list = 'snakemake_deduplicated/genome/reads/*.bam'\n",
    "    output:\n",
    "        counts = 'snakemake_counts/featurecounts/genome/featurecounts_genome.txt' \n",
    "    shell:\n",
    "        \"featureCounts -a {input.gtf_file} -o {output.counts} {params.reads_list}\"\n",
    "\n",
    "# clean columns in featurecounts matrix \n",
    "rule clean_featurecounts:\n",
    "    input:\n",
    "        'snakemake_counts/featurecounts/genome/featurecounts_genome.txt'\n",
    "    output:\n",
    "        'snakemake_counts/featurecounts/genome/featurecounts_genome_clean.txt'\n",
    "    script:\n",
    "        'scripts/clean_featurecounts.py'\n",
    "\n",
    "        \n",
    "# Testing different aligners and alignments\n",
    "\n",
    "# Did not align AT ALL!! \n",
    "rule bowtie2_mature_alignment:\n",
    "    input:\n",
    "        reads = 'snakemake_cutadapt/{file}_umi_cutadapt.fastq'\n",
    "    output:\n",
    "        aligned = 'snakemake_bowtie2/mature/{file}_mature_bowtie2_aligned.bam',\n",
    "        log = 'snakemake_bowtie2/mature/reports/{file}.log'\n",
    "    params:\n",
    "        index = 'indexes/bowtie2_index/mature/hsa_mature'\n",
    "    shell:\n",
    "        \"bowtie2 -k 100 --local --very-sensitive-local -x {params.index} -U {input.reads} 2> {output.log} | \" \n",
    "        \"samtools view -b - | samtools sort -o {output.aligned} - && samtools index {output.aligned}\"\n",
    "\n",
    "# trying salmon to mature.\n",
    "# maybe add --seqbias, --useVBOpt and --numBootStraps?\n",
    "rule salmon_mature_mirbase:\n",
    "    input:\n",
    "        'snakemake_cutadapt/{file}_umi_cutadapt.fastq'\n",
    "    output:\n",
    "        directory('snakemake_salmon/mature/{file}_salmon_mature/')\n",
    "    params:\n",
    "        index = 'indexes/salmon_index/mature/hsa_mature/'\n",
    "    shell:\n",
    "        \"\"\"\n",
    "        salmon quant -l A -i {params.index} -r {input} -o {output} --writeUnmappedNames -q\n",
    "        \"\"\"\n",
    "        \n",
    "# maybe add --seqbias, --useVBOpt(DEFAULT!) and --numBootStraps?        \n",
    "rule salmon_transcriptome:\n",
    "    input:\n",
    "        'snakemake_cutadapt/{file}_umi_cutadapt.fastq'\n",
    "    output:\n",
    "        directory('snakemake_salmon/transcriptome/{file}_salmon_transcriptome/')\n",
    "    params:\n",
    "        index = 'indexes/salmon_index/transcriptome/transcriptome/'\n",
    "    shell:\n",
    "        \"\"\"\n",
    "        salmon quant -l A --seqBias -i {params.index} -r {input} -o {output} --writeUnmappedNames -q\n",
    "        \"\"\"\n",
    "\n",
    "#testing bowtie2 against mirbase with T \n",
    "rule bowtie2_mature_T_alignment:\n",
    "    input:\n",
    "        reads = 'snakemake_cutadapt/{file}_umi_cutadapt.fastq'\n",
    "    output:\n",
    "        aligned = 'snakemake_bowtie2/mature_T/{file}_mature_bowtie2_aligned.bam',\n",
    "        log = 'snakemake_bowtie2/mature_T/reports/{file}.log'\n",
    "    params:\n",
    "        index = 'indexes/bowtie2_index/mature_T/hsa_mature_T'\n",
    "    shell:\n",
    "        \"bowtie2 -k 100 --local --very-sensitive-local -x {params.index} -U {input.reads} 2> {output.log} | \" \n",
    "        \"samtools view -b - | samtools sort -o {output.aligned} - && samtools index {output.aligned}\"\n",
    "        "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e0d83310-ef3f-4664-be31-0108e40be29b",
   "metadata": {},
   "outputs": [],
   "source": [
    "        "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "059883cc-2730-4978-87b3-2a827cd286f3",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Overwriting Snakefile\n"
     ]
    }
   ],
   "source": [
    "%%writefile Snakefile\n",
    "import subprocess\n",
    "from pathlib import Path\n",
    "\n",
    "\n",
    "# CHANGE ME\n",
    "READS = [str(file.stem).split('_')[0] for file in Path('snakemake_cutadapt/').iterdir()]\n",
    "\n",
    "\n",
    "rule all:\n",
    "    input:\n",
    "        expand('snakemake_star/{file}_aligned.sorted.bam', file = READS)\n",
    "        \n",
    "    \n",
    "rule remove_umi:\n",
    "    input:\n",
    "        raw_reads = 'test_data/{file}.gz'\n",
    "    output:\n",
    "        umi_reads = 'snakemake_umi/umi_trimmed/{file}_umi.gz',\n",
    "        log_file = 'snakemake_umi/reports/{file}_umi.log'\n",
    "    shell:\n",
    "        \"\"\"\n",
    "        umi_tools extract -p NNNNNNNN -I {input.raw_reads} -S {output.umi_reads} -L {output.log_file}\n",
    "        \"\"\"\n",
    "\n",
    "rule cutadapt:\n",
    "    input:\n",
    "        'snakemake_umi/umi_trimmed/{file}_umi.gz'\n",
    "    output:\n",
    "        cutadapt_reads = 'snakemake_cutadapt/{file}_umi_cutadapt.gz'\n",
    "    params:\n",
    "        adapter_file = 'adapters/adapters.txt'\n",
    "    shell:\n",
    "        \"\"\"\n",
    "        cutadapt -a file:{params.adapter_file} -e 0.1 -O 1 -u 2 --quiet --minimum-length 18 -o {output.cutadapt_reads} {input}\n",
    "        \"\"\"\n",
    "    \n",
    "# settings from Ziemann et al 2016: bowtie2 -k 100 --local --very-sensitive-local -x ref -U read\n",
    "# NOTE TO SELF: added samtools index here without testing\n",
    "# COMMENT: why not just use setting which dont allow soft clipping?? like end-to-end alignment???\n",
    "rule bowtie2_and_samtools:\n",
    "    input:\n",
    "        reads = 'snakemake_cutadapt/{file}_umi_cutadapt.gz'\n",
    "    output:\n",
    "        aligned_bam = 'snakemake_bowtie2/aligned/{file}_bowtie2_aligned.sorted.bam',\n",
    "        non_aligned = 'snakemake_bowtie2/not_aligned/{file}_bowtie2_non_aligned.gz',\n",
    "        log = 'snakemake_bowtie2/aligned/{file}_bowtie_log.txt'\n",
    "    params:\n",
    "        index = 'indexes/bowtie2_index/GRCh38'\n",
    "    shell:\n",
    "        \"\"\"\n",
    "        bowtie2 -x {params.index} -U {input.reads} --un-gz {output.non_aligned} 2> {output.log} | samtools view -b - | samtools sort -o {output.aligned_bam} - && samtools index {output.aligned_bam} \n",
    "        \"\"\"\n",
    "\n",
    "rule star:\n",
    "    input:\n",
    "        reads = 'snakemake_cutadapt/{file}_umi_cutadapt.gz'\n",
    "    output:\n",
    "        'snakemake_star/{file}_Aligned.out.sam',\n",
    "        'snakemake_star/{file}_Log.final.out',\n",
    "        'snakemake_star/{file}_Log.out',\n",
    "        'snakemake_star/{file}_Log.progress.out',\n",
    "        'snakemake_star/{file}_SJ.out.tab'\n",
    "    params:\n",
    "        index = 'indexes/star_index/',\n",
    "        prefix = 'snakemake_star/{file}_'\n",
    "    shell:\n",
    "        \"STAR --runThreadN 4 --genomeDir {params.index} --readFilesIn {input.reads} --readFilesCommand zcat --outSAMstrandField intronMotif \"\n",
    "        \"--outFilterMultimapNmax 50 --outFilterScoreMinOverLread 0 --outFilterMatchNmin 18 --outFilterMatchNminOverLread 0 --outFilterMismatchNoverLmax 0.04 --alignIntronMax 1 --outFileNamePrefix {params.prefix}\"\n",
    "\n",
    "# SHOUlD output sorted bam file from star instead!!!!\n",
    "rule sam_to_bam:\n",
    "    input:\n",
    "        sam = 'snakemake_star/{file}_Aligned.out.sam'\n",
    "    output:\n",
    "        sorted_bam = 'snakemake_star/{file}_aligned.sorted.bam'\n",
    "    shell:\n",
    "        \"\"\"\n",
    "        samtools view -bS {input.sam} | samtools sort -o {output.sorted_bam} - && samtools index {output.sorted_bam}\n",
    "        \"\"\"\n",
    "\n",
    "#add script here \n",
    "rule remove_softclip:\n",
    "    input:\n",
    "        'snakemake_star/{file}_aligned.sorted.bam' #maybe input sam? \n",
    "    output:\n",
    "        'removed_clipped/{file}_clipped.bam'\n",
    "    script:\n",
    "        'scripts/XXXX.py'\n",
    "\n",
    "\n",
    "rule subsample_bam_readlength:\n",
    "    input:\n",
    "        'snakemake_star/{file}_aligned.sorted.bam'\n",
    "    output:\n",
    "        'snakemake_subsampled/{file}_subsampled.bam'\n",
    "    params:\n",
    "        max_cutoff = 40 \n",
    "        \n",
    "    shell:\n",
    "        \"\"\"\n",
    "        samtools view -h {input} | awk \\'length($10) <= {params.max_cutoff} || $1~\"@\"\\' | samtools view -bS - > {output}\n",
    "        \"\"\"\n",
    "\n",
    "rule umi_deduplication:\n",
    "    input:\n",
    "        'snakemake_subsampled/{file}_subsampled.bam'\n",
    "    output:\n",
    "        deduplicated = 'snakemake_removed_umi/{file}_removedumi.bam',\n",
    "        log = 'snakemake_removed_umi/{file}_log.txt'\n",
    "    params:\n",
    "        output_stats = {file}\n",
    "    shell:\n",
    "        \"\"\"\n",
    "        umi_tools dedup --method adjacency --output-stats={params.output_stats} --further-stats --read-length -I {input} -S {output.deduplicated} -L {output.log}\n",
    "        \"\"\"\n",
    "#rule index_bam:\n",
    "#rule merge_bam:\n",
    "#rule fastqc:\n",
    "#rule multiqc:  \n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e2cb94e9-43fa-4fce-bfc5-c0c45608c10a",
   "metadata": {},
   "source": [
    "# SCRIPTS"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "a4322013-7844-4881-a992-5b30c566cfbc",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Overwriting scripts/download_SRA.py\n"
     ]
    }
   ],
   "source": [
    "%%writefile scripts/download_SRA.py\n",
    "import pandas as pd\n",
    "import subprocess\n",
    "\n",
    "sra = pd.read_csv('sra_table.txt')\n",
    "\n",
    "for file in sra.Run:\n",
    "    command = f\"prefetch {file}\".split()\n",
    "    subprocess.call(command)\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a31a439f-e6c7-49ca-98d2-bbde7044c050",
   "metadata": {},
   "source": [
    "# TESTCODE"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "b3a763bf-1924-4b62-8389-037e90949c4b",
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    },
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>run</th>\n",
       "      <th>assay_type</th>\n",
       "      <th>avgspotlen</th>\n",
       "      <th>bases</th>\n",
       "      <th>bioproject</th>\n",
       "      <th>biosample</th>\n",
       "      <th>bytes</th>\n",
       "      <th>center_name</th>\n",
       "      <th>consent</th>\n",
       "      <th>datastore_filetype</th>\n",
       "      <th>...</th>\n",
       "      <th>librarylayout</th>\n",
       "      <th>libraryselection</th>\n",
       "      <th>librarysource</th>\n",
       "      <th>organism</th>\n",
       "      <th>platform</th>\n",
       "      <th>releasedate</th>\n",
       "      <th>sample_name</th>\n",
       "      <th>source_name</th>\n",
       "      <th>sra_study</th>\n",
       "      <th>cell_type</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th style=\"font-style: italic;\" ></th>\n",
       "      <td style=\"font-style: italic;\" >&lt;object&gt;</td>\n",
       "      <td style=\"font-style: italic;\" >&lt;object&gt;</td>\n",
       "      <td style=\"font-style: italic;\" >&lt;int64&gt;</td>\n",
       "      <td style=\"font-style: italic;\" >&lt;int64&gt;</td>\n",
       "      <td style=\"font-style: italic;\" >&lt;object&gt;</td>\n",
       "      <td style=\"font-style: italic;\" >&lt;object&gt;</td>\n",
       "      <td style=\"font-style: italic;\" >&lt;int64&gt;</td>\n",
       "      <td style=\"font-style: italic;\" >&lt;object&gt;</td>\n",
       "      <td style=\"font-style: italic;\" >&lt;object&gt;</td>\n",
       "      <td style=\"font-style: italic;\" >&lt;object&gt;</td>\n",
       "      <td style=\"font-style: italic;\" >...</td>\n",
       "      <td style=\"font-style: italic;\" >&lt;object&gt;</td>\n",
       "      <td style=\"font-style: italic;\" >&lt;object&gt;</td>\n",
       "      <td style=\"font-style: italic;\" >&lt;object&gt;</td>\n",
       "      <td style=\"font-style: italic;\" >&lt;object&gt;</td>\n",
       "      <td style=\"font-style: italic;\" >&lt;object&gt;</td>\n",
       "      <td style=\"font-style: italic;\" >&lt;object&gt;</td>\n",
       "      <td style=\"font-style: italic;\" >&lt;object&gt;</td>\n",
       "      <td style=\"font-style: italic;\" >&lt;object&gt;</td>\n",
       "      <td style=\"font-style: italic;\" >&lt;object&gt;</td>\n",
       "      <td style=\"font-style: italic;\" >&lt;object&gt;</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>SRR3495523</td>\n",
       "      <td>ncRNA-Seq</td>\n",
       "      <td>43</td>\n",
       "      <td>110906503</td>\n",
       "      <td>PRJNA321192</td>\n",
       "      <td>SAMN04976967</td>\n",
       "      <td>112821441</td>\n",
       "      <td>GEO</td>\n",
       "      <td>public</td>\n",
       "      <td>fastq,sra</td>\n",
       "      <td>...</td>\n",
       "      <td>SINGLE</td>\n",
       "      <td>size fractionation</td>\n",
       "      <td>TRANSCRIPTOMIC</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>ILLUMINA</td>\n",
       "      <td>2016-10-17T00:00:00Z</td>\n",
       "      <td>GSM2149376</td>\n",
       "      <td>Glioblastoma</td>\n",
       "      <td>SRP074776</td>\n",
       "      <td>GBM</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>SRR3495524</td>\n",
       "      <td>ncRNA-Seq</td>\n",
       "      <td>43</td>\n",
       "      <td>108781615</td>\n",
       "      <td>PRJNA321192</td>\n",
       "      <td>SAMN04976968</td>\n",
       "      <td>109873177</td>\n",
       "      <td>GEO</td>\n",
       "      <td>public</td>\n",
       "      <td>fastq,sra</td>\n",
       "      <td>...</td>\n",
       "      <td>SINGLE</td>\n",
       "      <td>size fractionation</td>\n",
       "      <td>TRANSCRIPTOMIC</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>ILLUMINA</td>\n",
       "      <td>2016-10-17T00:00:00Z</td>\n",
       "      <td>GSM2149377</td>\n",
       "      <td>Glioblastoma</td>\n",
       "      <td>SRP074776</td>\n",
       "      <td>GBM</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>SRR3495525</td>\n",
       "      <td>ncRNA-Seq</td>\n",
       "      <td>43</td>\n",
       "      <td>149699383</td>\n",
       "      <td>PRJNA321192</td>\n",
       "      <td>SAMN04976969</td>\n",
       "      <td>150617486</td>\n",
       "      <td>GEO</td>\n",
       "      <td>public</td>\n",
       "      <td>fastq,sra</td>\n",
       "      <td>...</td>\n",
       "      <td>SINGLE</td>\n",
       "      <td>size fractionation</td>\n",
       "      <td>TRANSCRIPTOMIC</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>ILLUMINA</td>\n",
       "      <td>2016-10-17T00:00:00Z</td>\n",
       "      <td>GSM2149378</td>\n",
       "      <td>Glioblastoma</td>\n",
       "      <td>SRP074776</td>\n",
       "      <td>GBM</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>SRR3495526</td>\n",
       "      <td>ncRNA-Seq</td>\n",
       "      <td>43</td>\n",
       "      <td>157402618</td>\n",
       "      <td>PRJNA321192</td>\n",
       "      <td>SAMN04976970</td>\n",
       "      <td>158483005</td>\n",
       "      <td>GEO</td>\n",
       "      <td>public</td>\n",
       "      <td>fastq,sra</td>\n",
       "      <td>...</td>\n",
       "      <td>SINGLE</td>\n",
       "      <td>size fractionation</td>\n",
       "      <td>TRANSCRIPTOMIC</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>ILLUMINA</td>\n",
       "      <td>2016-10-17T00:00:00Z</td>\n",
       "      <td>GSM2149379</td>\n",
       "      <td>Glioblastoma</td>\n",
       "      <td>SRP074776</td>\n",
       "      <td>GBM</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>SRR3495527</td>\n",
       "      <td>ncRNA-Seq</td>\n",
       "      <td>43</td>\n",
       "      <td>129724206</td>\n",
       "      <td>PRJNA321192</td>\n",
       "      <td>SAMN04976971</td>\n",
       "      <td>131526400</td>\n",
       "      <td>GEO</td>\n",
       "      <td>public</td>\n",
       "      <td>fastq,sra</td>\n",
       "      <td>...</td>\n",
       "      <td>SINGLE</td>\n",
       "      <td>size fractionation</td>\n",
       "      <td>TRANSCRIPTOMIC</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>ILLUMINA</td>\n",
       "      <td>2016-10-17T00:00:00Z</td>\n",
       "      <td>GSM2149380</td>\n",
       "      <td>Glioblastoma</td>\n",
       "      <td>SRP074776</td>\n",
       "      <td>GBM</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>SRR3495528</td>\n",
       "      <td>ncRNA-Seq</td>\n",
       "      <td>43</td>\n",
       "      <td>133369359</td>\n",
       "      <td>PRJNA321192</td>\n",
       "      <td>SAMN04976972</td>\n",
       "      <td>135118332</td>\n",
       "      <td>GEO</td>\n",
       "      <td>public</td>\n",
       "      <td>fastq,sra</td>\n",
       "      <td>...</td>\n",
       "      <td>SINGLE</td>\n",
       "      <td>size fractionation</td>\n",
       "      <td>TRANSCRIPTOMIC</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>ILLUMINA</td>\n",
       "      <td>2016-10-17T00:00:00Z</td>\n",
       "      <td>GSM2149381</td>\n",
       "      <td>Glioblastoma</td>\n",
       "      <td>SRP074776</td>\n",
       "      <td>GBM</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>6 rows Ã— 25 columns</p>"
      ],
      "text/plain": [
       "          run assay_type  avgspotlen      bases   bioproject     biosample  \\\n",
       "     <object>   <object>     <int64>    <int64>     <object>      <object>   \n",
       "0  SRR3495523  ncRNA-Seq          43  110906503  PRJNA321192  SAMN04976967   \n",
       "1  SRR3495524  ncRNA-Seq          43  108781615  PRJNA321192  SAMN04976968   \n",
       "2  SRR3495525  ncRNA-Seq          43  149699383  PRJNA321192  SAMN04976969   \n",
       "3  SRR3495526  ncRNA-Seq          43  157402618  PRJNA321192  SAMN04976970   \n",
       "4  SRR3495527  ncRNA-Seq          43  129724206  PRJNA321192  SAMN04976971   \n",
       "5  SRR3495528  ncRNA-Seq          43  133369359  PRJNA321192  SAMN04976972   \n",
       "\n",
       "       bytes center_name  consent datastore_filetype  ... librarylayout  \\\n",
       "     <int64>    <object> <object>           <object>  ...      <object>   \n",
       "0  112821441         GEO   public          fastq,sra  ...        SINGLE   \n",
       "1  109873177         GEO   public          fastq,sra  ...        SINGLE   \n",
       "2  150617486         GEO   public          fastq,sra  ...        SINGLE   \n",
       "3  158483005         GEO   public          fastq,sra  ...        SINGLE   \n",
       "4  131526400         GEO   public          fastq,sra  ...        SINGLE   \n",
       "5  135118332         GEO   public          fastq,sra             SINGLE   \n",
       "\n",
       "     libraryselection   librarysource      organism  platform  \\\n",
       "             <object>        <object>      <object>  <object>   \n",
       "0  size fractionation  TRANSCRIPTOMIC  Homo sapiens  ILLUMINA   \n",
       "1  size fractionation  TRANSCRIPTOMIC  Homo sapiens  ILLUMINA   \n",
       "2  size fractionation  TRANSCRIPTOMIC  Homo sapiens  ILLUMINA   \n",
       "3  size fractionation  TRANSCRIPTOMIC  Homo sapiens  ILLUMINA   \n",
       "4  size fractionation  TRANSCRIPTOMIC  Homo sapiens  ILLUMINA   \n",
       "5  size fractionation  TRANSCRIPTOMIC  Homo sapiens  ILLUMINA   \n",
       "\n",
       "            releasedate sample_name   source_name  sra_study cell_type  \n",
       "               <object>    <object>      <object>   <object>  <object>  \n",
       "0  2016-10-17T00:00:00Z  GSM2149376  Glioblastoma  SRP074776       GBM  \n",
       "1  2016-10-17T00:00:00Z  GSM2149377  Glioblastoma  SRP074776       GBM  \n",
       "2  2016-10-17T00:00:00Z  GSM2149378  Glioblastoma  SRP074776       GBM  \n",
       "3  2016-10-17T00:00:00Z  GSM2149379  Glioblastoma  SRP074776       GBM  \n",
       "4  2016-10-17T00:00:00Z  GSM2149380  Glioblastoma  SRP074776       GBM  \n",
       "5  2016-10-17T00:00:00Z  GSM2149381  Glioblastoma  SRP074776       GBM  \n",
       "\n",
       "[6 rows x 25 columns]"
      ]
     },
     "execution_count": 29,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import pandas as pd\n",
    "from janitor import clean_names\n",
    "from datar.all import *\n",
    "\n",
    "sra = pd.read_csv('test_data/sra_table.txt').clean_names() >> select(~f.sample_type)\n",
    "\n",
    "sra >> filter(f.cell_type == 'GBM')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "id": "313a7ce0-71a0-4052-a100-52b48db8c714",
   "metadata": {},
   "outputs": [],
   "source": [
    "from pathlib import Path\n",
    "\n",
    "files = [file for file in Path('snakemake_cutadapt/').iterdir()]\n",
    "\n",
    "for file in files:\n",
    "    new = str(file).replace('.fastq', '')\n",
    "    file.rename(new)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5088798d-e539-41cc-9a5b-00caf9810e32",
   "metadata": {
    "tags": []
   },
   "source": [
    "# SNAKEMAKE SCRIPTS"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "be207c95-a35a-4866-87f9-57db26edd427",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Overwriting scripts/remove_soft_clip.py\n"
     ]
    }
   ],
   "source": [
    "%%writefile scripts/remove_soft_clip.py\n",
    "\n",
    "# TODO: Fix this to insert snakemake.inputs and outputs. PRobably dont have to loop through, since it will pipe one sample each go. \n",
    "\n",
    "\n",
    "import pysam\n",
    "from pathlib import Path\n",
    "\n",
    "\n",
    "def remove_reads(in_bam_file):\n",
    "    in_bam = pysam.AlignmentFile(in_bam_file)\n",
    "    path_out_bam = Path(out_dir, (in_bam_file.stem + 'clipped.bam'))\n",
    "    out_bam = pysam.AlignmentFile(path_out_bam, 'wb', template=in_bam)\n",
    "    total_reads = 0\n",
    "    reads_after_clipped = 0\n",
    "    logfile = Path(out_dir, 'logfile.txt')\n",
    "    \n",
    "    #limits\n",
    "    allowed_num_clipped_bases_5p = 0 #snakemake.params.variable ?? \n",
    "    allowed_num_clipped_bases_3p = 3\n",
    "        \n",
    "    for read in in_bam:\n",
    "        total_reads += 1\n",
    "        read_name = read.qname\n",
    "        cigar = read.cigar  #this is a list of tuples;  example: cigar=[(0, 25), (4, 8)]\n",
    "        \n",
    "        num_softclipped_5p = 0\n",
    "        num_softclipped_3p = 0\n",
    "        num_hardclipped = 0\n",
    "        num_insertions = 0\n",
    "        num_deletions = 0\n",
    "        \n",
    "        for i,e in enumerate(cigar):\n",
    "            if e[0] == 5:\n",
    "                num_hardclipped += e[1]\n",
    "            elif e[0] == 1:\n",
    "                num_insertions += e[1]\n",
    "            elif e[0] == 2:\n",
    "                num_deletions += e[1]\n",
    "            elif i == 0 and e[0] == 4:\n",
    "                num_softclipped_5p += e[1]\n",
    "            elif i == len(cigar)-1 and e[0] == 4:\n",
    "                num_softclipped_3p += e[1]\n",
    "\n",
    "        if num_softclipped_5p > allowed_num_clipped_bases_5p:\n",
    "            continue\n",
    "        elif num_softclipped_3p > allowed_num_clipped_bases_3p:\n",
    "            continue\n",
    "        elif num_hardclipped > 0: \n",
    "            continue\n",
    "        elif num_insertions > 0: \n",
    "            continue \n",
    "        elif num_deletions > 0: \n",
    "            continue   \n",
    "        \n",
    "        out_bam.write(read)\n",
    "        reads_after_clipped += 1\n",
    "        \n",
    "    out_bam.close()\n",
    "    \n",
    "    pysam.sort('-o', str(path_out_bam), str(path_out_bam))  \n",
    "    pysam.index(str(path_out_bam))\n",
    "    \n",
    "    with open(logfile, 'a') as f:\n",
    "        print(f\" Procent removed in {read_name}: {100 * (1- reads_after_clipped / total_reads): .2f}%\", file=f)\n",
    "\n",
    "\n",
    "        \n",
    "#directory making\n",
    "out_dir = Path('clipped_removed/') #snakemake.output\n",
    "if not out_dir.exists():\n",
    "    out_dir.mkdir(parents=True, exist_ok=True)\n",
    "\n",
    "    \n",
    "files = [file for file in Path('snakemake_star/').iterdir() if file.suffix == '.bam'] #snakemake.input\n",
    "        \n",
    "#main loop\n",
    "for bamfile in files:\n",
    "    print(f'started {bamfile}')\n",
    "    remove_reads(bamfile)\n",
    "    print(f'done {bamfile}')\n",
    "  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 141,
   "id": "361aafc6-497c-4787-8abe-2c10caf7f141",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Overwriting scripts/clean_featurecounts.py\n"
     ]
    }
   ],
   "source": [
    "%%writefile scripts/clean_featurecounts.py\n",
    "import pandas as pd\n",
    "from datar.all import select, contains\n",
    "import re\n",
    "import os\n",
    "\n",
    "\n",
    "# read in data \n",
    "in_file = snakemake.input[0]\n",
    "out_file = snakemake.output[0]\n",
    "counts = pd.read_csv(in_file, delimiter='\\t', skiprows=1)\n",
    "\n",
    "# filter for correct columns\n",
    "counts = (counts >>\n",
    " select(contains('Geneid') | contains('bam'))\n",
    ")\n",
    "\n",
    "# extract SRR number from the columns\n",
    "pattern = re.compile(r'SRR\\d+')\n",
    "new_names = [\"\".join(re.findall(pattern, column)) if not column == 'Geneid' else column for column in counts.columns]\n",
    "\n",
    "# rename columns with SRR number and save file \n",
    "counts.columns = new_names\n",
    "counts.to_csv(out_file)\n",
    "\n",
    "# remove original file\n",
    "os.remove(in_file)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 174,
   "id": "693a6d5e-e63b-4c68-b991-3bad76ada593",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Writing snakefile_star_debo\n"
     ]
    }
   ],
   "source": [
    "%%writefile snakefile_star_debo\n",
    "\n",
    "## new snake file to only run star to show debo\n",
    "\n",
    "\n",
    "READS = [file.stem for file in Path('SRR_new/').iterdir() if file.suffix == '.fastq']\n",
    "\n",
    "rule all:\n",
    "    input:\n",
    "        expand('snakemake_star/{file}_aligned.sorted.bam', file = READS)\n",
    "        \n",
    "    \n",
    "rule remove_umi:\n",
    "    input:\n",
    "        raw_reads = 'SRR_new/{file}.fastq'\n",
    "    output:\n",
    "        umi_reads = 'snakemake_umi/umi_trimmed/{file}_umi.fastq',\n",
    "        log_file = 'snakemake_umi/reports/{file}_umi.log'\n",
    "    shell:\n",
    "        \"umi_tools extract --extract-method=regex --bc-pattern='(?P<discard_1>.*)(?P<umi_1>[ACT]{{8}}CA)' \" \n",
    "        \"-I {input.raw_reads} -S {output.umi_reads} -L {output.log_file}\" \n",
    "        #FOR REMOVING 3' ADAPTER: '(?P<discard_1>.*)(?P<umi_1>[ACT]{{8}}CA).*(?P<discard_2>TGGAATTCT.+)' \n",
    "        \n",
    "# REMOVED -u here because it filters the 5' ends of the read. I have already done this in umi_tools! \n",
    "rule cutadapt:\n",
    "    input:\n",
    "        'snakemake_umi/umi_trimmed/{file}_umi.fastq'\n",
    "    output:\n",
    "        cutadapt_reads = 'snakemake_cutadapt/{file}_umi_cutadapt.fastq',\n",
    "        log = 'snakemake_cutadapt/reports/{file}_cutadapt.log'\n",
    "    params:\n",
    "        adapter_file = 'adapters/adapters.txt'\n",
    "    shell:\n",
    "        \"cutadapt -a file:{params.adapter_file} -e 0.1 -O 1 --quiet --minimum-length 18 \" \n",
    "        \"--maximum-length 40 -o {output.cutadapt_reads} {input} > {output.log}\"\n",
    "\n",
    "rule star_all_reads:\n",
    "    input:\n",
    "        reads = 'snakemake_cutadapt/{file}_umi_cutadapt.fastq'\n",
    "    output:\n",
    "        'snakemake_star/{file}_Aligned.out.sam',\n",
    "        'snakemake_star/{file}_Log.final.out',\n",
    "        'snakemake_star/{file}_Log.out',\n",
    "        'snakemake_star/{file}_Log.progress.out',\n",
    "        'snakemake_star/{file}_SJ.out.tab'\n",
    "    params:\n",
    "        index = 'indexes/star_index/',\n",
    "        prefix = 'snakemake_star/{file}_'\n",
    "    shell:\n",
    "        \"STAR --runThreadN 4 --genomeDir {params.index} --readFilesIn {input.reads} \" \n",
    "        \"--readFilesCommand zcat --outSAMstrandField intronMotif \"\n",
    "        \"--outFilterMultimapNmax 50 --outFilterScoreMinOverLread 0 --outFilterMatchNmin 18 \" \n",
    "        \"--outFilterMatchNminOverLread 0 --outFilterMismatchNoverLmax 0.04 --alignIntronMax 1 --outFileNamePrefix {params.prefix}\"\n",
    "\n",
    "rule star_sam_to_bam:\n",
    "    input:\n",
    "        aligned_sam = 'snakemake_star/{file}_Aligned.out.sam'\n",
    "    output:\n",
    "        sorted_bam = 'snakemake_star/{file}_aligned.sorted.bam'\n",
    "    shell:\n",
    "        \"samtools sort -o {output.sorted_bam} {input.aligned_sam} && rm {input.aligned_sam} \"\n",
    "        \"&& samtools index {output.sorted_bam}\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 179,
   "id": "657bfb0b-ea79-4473-9f46-afd8abc58476",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Overwriting snakefile_star_debo_mildare_parametrar\n"
     ]
    }
   ],
   "source": [
    "%%writefile snakefile_star_debo_mildare_parametrar\n",
    "\n",
    "## new snake file to only run star to show debo, changed the cutadapt parameters to include a broader range of reads. \n",
    "## This did NOT align at all! \n",
    "\n",
    "\n",
    "READS = [file.stem for file in Path('SRR_new/').iterdir() if file.suffix == '.fastq']\n",
    "\n",
    "rule all:\n",
    "    input:\n",
    "        expand('snakemake_star/{file}_aligned.sorted.bam', file = READS)\n",
    "        \n",
    "    \n",
    "rule remove_umi:\n",
    "    input:\n",
    "        raw_reads = 'SRR_new/{file}.fastq'\n",
    "    output:\n",
    "        umi_reads = 'snakemake_umi/umi_trimmed/{file}_umi.fastq',\n",
    "        log_file = 'snakemake_umi/reports/{file}_umi.log'\n",
    "    shell:\n",
    "        \"umi_tools extract --extract-method=regex --bc-pattern='(?P<discard_1>.*)(?P<umi_1>[ACT]{{8}}CA)' \" \n",
    "        \"-I {input.raw_reads} -S {output.umi_reads} -L {output.log_file}\" \n",
    "        #FOR REMOVING 3' ADAPTER: '(?P<discard_1>.*)(?P<umi_1>[ACT]{{8}}CA).*(?P<discard_2>TGGAATTCT.+)' \n",
    "        \n",
    "# REMOVED -u here because it filters the 5' ends of the read. I have already done this in umi_tools! \n",
    "rule cutadapt:\n",
    "    input:\n",
    "        'snakemake_umi/umi_trimmed/{file}_umi.fastq'\n",
    "    output:\n",
    "        cutadapt_reads = 'snakemake_cutadapt/{file}_umi_cutadapt.fastq',\n",
    "        log = 'snakemake_cutadapt/reports/{file}_cutadapt.log'\n",
    "    params:\n",
    "        adapter_file = 'adapters/adapters.txt'\n",
    "    shell:\n",
    "        \"cutadapt -a file:{params.adapter_file} -e 0.1 -O 1 --quiet --minimum-length 12 \" \n",
    "        \"--maximum-length 50 -o {output.cutadapt_reads} {input} > {output.log}\"\n",
    "\n",
    "rule star_all_reads:\n",
    "    input:\n",
    "        reads = 'snakemake_cutadapt/{file}_umi_cutadapt.fastq'\n",
    "    output:\n",
    "        'snakemake_star/{file}_Aligned.out.sam',\n",
    "        'snakemake_star/{file}_Log.final.out',\n",
    "        'snakemake_star/{file}_Log.out',\n",
    "        'snakemake_star/{file}_Log.progress.out',\n",
    "        'snakemake_star/{file}_SJ.out.tab'\n",
    "    params:\n",
    "        index = 'indexes/star_index_ensembl/',\n",
    "        prefix = 'snakemake_star/{file}_'\n",
    "    shell:\n",
    "        \"STAR --runThreadN 4 --genomeDir {params.index} --readFilesIn {input.reads} \" \n",
    "        \"--readFilesCommand zcat --outSAMstrandField intronMotif \"\n",
    "        \"--outFilterMultimapNmax 50 --outFilterScoreMinOverLread 0 --outFilterMatchNmin 12 \" \n",
    "        \"--outFilterMatchNminOverLread 0 --outFilterMismatchNoverLmax 0.04 --alignIntronMax 1 --outFileNamePrefix {params.prefix}\"\n",
    "\n",
    "rule star_sam_to_bam:\n",
    "    input:\n",
    "        aligned_sam = 'snakemake_star/{file}_Aligned.out.sam'\n",
    "    output:\n",
    "        sorted_bam = 'snakemake_star/{file}_aligned.sorted.bam'\n",
    "    shell:\n",
    "        \"samtools sort -o {output.sorted_bam} {input.aligned_sam} && rm {input.aligned_sam} \"\n",
    "        \"&& samtools index {output.sorted_bam}\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 191,
   "id": "8e8cfc48-d52b-41b2-b077-829b743810ed",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Overwriting snakefile_different_STAR_parameters\n"
     ]
    }
   ],
   "source": [
    "%%writefile snakefile_different_STAR_parameters\n",
    "\n",
    "## STAR parameters from ENCODE miRNA-seq read alignment using STAR aligner,\n",
    "#https://www.encodeproject.org/documents/b4ec4567-ac4e-4812-b2bd-e1d2df746966/@@download/attachment/ENCODE_miRNA-seq_STAR_parameters_v2.pdf\n",
    "\n",
    "READS = [file.stem for file in Path('SUBSET_DATA/').iterdir() if file.suffix == '.fastq']\n",
    "\n",
    "rule all:\n",
    "    input:\n",
    "        expand('snakemake_star/{file}_Aligned.sortedByCoord.out.bam', file = READS)\n",
    "        \n",
    "    \n",
    "rule remove_umi:\n",
    "    input:\n",
    "        raw_reads = 'SUBSET_DATA/{file}.fastq'\n",
    "    output:\n",
    "        umi_reads = 'snakemake_umi/umi_trimmed/{file}_umi.fastq',\n",
    "        log_file = 'snakemake_umi/reports/{file}_umi.log'\n",
    "    shell:\n",
    "        \"umi_tools extract --extract-method=regex --bc-pattern='(?P<discard_1>.*)(?P<umi_1>[ACT]{{8}}CA)' \" \n",
    "        \"-I {input.raw_reads} -S {output.umi_reads} -L {output.log_file}\" \n",
    "        #FOR REMOVING 3' ADAPTER: '(?P<discard_1>.*)(?P<umi_1>[ACT]{{8}}CA).*(?P<discard_2>TGGAATTCT.+)' \n",
    "        \n",
    "# REMOVED -u here because it filters the 5' ends of the read. I have already done this in umi_tools! \n",
    "rule cutadapt:\n",
    "    input:\n",
    "        'snakemake_umi/umi_trimmed/{file}_umi.fastq'\n",
    "    output:\n",
    "        cutadapt_reads = 'snakemake_cutadapt/{file}_umi_cutadapt.fastq',\n",
    "        log = 'snakemake_cutadapt/reports/{file}_cutadapt.log'\n",
    "    params:\n",
    "        adapter_file = 'adapters/adapters.txt'\n",
    "    shell:\n",
    "        \"cutadapt -a file:{params.adapter_file} -e 0.1 -O 1 --quiet --minimum-length 18 \" \n",
    "        \"--maximum-length 40 -o {output.cutadapt_reads} {input} > {output.log}\"\n",
    "\n",
    "rule star_all_reads:\n",
    "    input:\n",
    "        reads = 'snakemake_cutadapt/{file}_umi_cutadapt.fastq',\n",
    "        gtf = 'genomes/gencode_miRNA_subset.gtf'\n",
    "    output:\n",
    "        'snakemake_star/{file}_Aligned.sortedByCoord.out.bam'\n",
    "    params:\n",
    "        index = 'indexes/star_index_gencode/',\n",
    "        prefix = 'snakemake_star/{file}_'\n",
    "    shell:\n",
    "        \"STAR --runThreadN 4 \"\n",
    "        \"--genomeDir {params.index} --readFilesIn {input.reads} \"\n",
    "        \"--sjdbGTFfile {input.gtf} \"\n",
    "        \"--outFileNamePrefix {params.prefix} \"\n",
    "        \"--alignEndsType EndToEnd \"\n",
    "        \"--outFilterMismatchNmax 1 \"\n",
    "        \"--outFilterMultimapScoreRange 0 \"\n",
    "        \"--quantMode TranscriptomeSAM GeneCounts \"\n",
    "        \"--outReadsUnmapped Fastx \"\n",
    "        \"--outSAMtype BAM SortedByCoordinate \"\n",
    "        \"--outFilterMultimapNmax 10 \"\n",
    "        \"--outSAMunmapped Within \"\n",
    "        \"--outFilterScoreMinOverLread 0 \"\n",
    "        \"--outFilterMatchNmin 16 \"\n",
    "        \"--alignSJDBoverhangMin 1000 \"\n",
    "        \"--alignIntronMax 1 \"\n",
    "        \"--outWigType wiggle \"\n",
    "        \"--outWigStrand Stranded \"\n",
    "        \"--outWigNorm RPM \"\n",
    "\n",
    "\n",
    "    \n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 195,
   "id": "0e9f5555-c30d-4cac-b0a0-20e31cb3a608",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Overwriting snakefile_bowtie1_different_parameters\n"
     ]
    }
   ],
   "source": [
    "%%writefile snakefile_bowtie1_different_parameters\n",
    "\n",
    "## Testing Alignment with bowtie with parameters found in \n",
    "# https://www.biostars.org/p/360709/\n",
    "# This aligns well (73% to the genome), however, should make it a little bit more stringent!! \n",
    "\n",
    "READS = [file.stem for file in Path('SUBSET_DATA/').iterdir() if file.suffix == '.fastq']\n",
    "\n",
    "rule all:\n",
    "    input:\n",
    "        expand('snakemake_bowtie/aligned/{file}_bowtie_aligned.sam', file = READS)\n",
    "        \n",
    "    \n",
    "rule remove_umi:\n",
    "    input:\n",
    "        raw_reads = 'SUBSET_DATA/{file}.fastq'\n",
    "    output:\n",
    "        umi_reads = 'snakemake_umi/umi_trimmed/{file}_umi.fastq',\n",
    "        log_file = 'snakemake_umi/reports/{file}_umi.log'\n",
    "    shell:\n",
    "        \"umi_tools extract --extract-method=regex --bc-pattern='(?P<discard_1>.*)(?P<umi_1>[ACT]{{8}}CA)' \" \n",
    "        \"-I {input.raw_reads} -S {output.umi_reads} -L {output.log_file}\" \n",
    "        #FOR REMOVING 3' ADAPTER: '(?P<discard_1>.*)(?P<umi_1>[ACT]{{8}}CA).*(?P<discard_2>TGGAATTCT.+)' \n",
    "        \n",
    "rule cutadapt:\n",
    "    input:\n",
    "        'snakemake_umi/umi_trimmed/{file}_umi.fastq'\n",
    "    output:\n",
    "        cutadapt_reads = 'snakemake_cutadapt/{file}_umi_cutadapt.fastq',\n",
    "        log = 'snakemake_cutadapt/reports/{file}_cutadapt.log'\n",
    "    params:\n",
    "        adapter_file = 'adapters/adapters.txt'\n",
    "    shell:\n",
    "        \"cutadapt -a file:{params.adapter_file} -e 0.1 -O 1 --quiet --minimum-length 18 \" \n",
    "        \"--maximum-length 40 -o {output.cutadapt_reads} {input} > {output.log}\"\n",
    "\n",
    "\n",
    "rule bowtie1_alignment:\n",
    "    input:\n",
    "        reads = 'snakemake_cutadapt/{file}_umi_cutadapt.fastq'\n",
    "    output:\n",
    "        aligned_sam = 'snakemake_bowtie/aligned/{file}_bowtie_aligned.sam',\n",
    "        non_aligned = 'snakemake_bowtie/not_aligned/{file}_bowtie_non_aligned.fastq',\n",
    "        log = 'snakemake_bowtie/reports/{file}_bowtie_aligned.log'\n",
    "    params:\n",
    "        index = 'indexes/bowtie_index/hg_38_index/hg_38'\n",
    "    shell:\n",
    "        \"bowtie -n 1 -l 10 --best --strata -m 100 -k 1 -x {params.index} {input.reads} \"\n",
    "        \"--un {output.non_aligned} -S {output.aligned_sam} 2> {output.log}\" \n",
    "\n",
    "#bowtie -n 0 -l 32 --norc --best --strata -m 1 (From the previous tries)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 199,
   "id": "31517387-acbe-4995-bb8c-9d3538b7ff0a",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Overwriting snakefile_bowtie1_stringent_parameters\n"
     ]
    }
   ],
   "source": [
    "%%writefile snakefile_bowtie1_stringent_parameters\n",
    "\n",
    "## Testing Alignment with bowtie with parameters found in \n",
    "# https://www.biostars.org/p/360709/\n",
    "# Changing the -m to 1 and -n to 0 to be more stringent\n",
    "# --> leads to much more stringent alignment, and it filters out much more reads (around 43% vs 1% due to low m)\n",
    "\n",
    "\n",
    "READS = [file.stem for file in Path('SUBSET_DATA/').iterdir() if file.suffix == '.fastq']\n",
    "\n",
    "rule all:\n",
    "    input:\n",
    "        expand('snakemake_bowtie_stringent/aligned/{file}_bowtie_aligned.sam', file = READS)\n",
    "        \n",
    "    \n",
    "rule remove_umi:\n",
    "    input:\n",
    "        raw_reads = 'SUBSET_DATA/{file}.fastq'\n",
    "    output:\n",
    "        umi_reads = 'snakemake_umi/umi_trimmed/{file}_umi.fastq',\n",
    "        log_file = 'snakemake_umi/reports/{file}_umi.log'\n",
    "    shell:\n",
    "        \"umi_tools extract --extract-method=regex --bc-pattern='(?P<discard_1>.*)(?P<umi_1>[ACT]{{8}}CA)' \" \n",
    "        \"-I {input.raw_reads} -S {output.umi_reads} -L {output.log_file}\" \n",
    "        #FOR REMOVING 3' ADAPTER: '(?P<discard_1>.*)(?P<umi_1>[ACT]{{8}}CA).*(?P<discard_2>TGGAATTCT.+)' \n",
    "        \n",
    "rule cutadapt:\n",
    "    input:\n",
    "        'snakemake_umi/umi_trimmed/{file}_umi.fastq'\n",
    "    output:\n",
    "        cutadapt_reads = 'snakemake_cutadapt/{file}_umi_cutadapt.fastq',\n",
    "        log = 'snakemake_cutadapt/reports/{file}_cutadapt.log'\n",
    "    params:\n",
    "        adapter_file = 'adapters/adapters.txt'\n",
    "    shell:\n",
    "        \"cutadapt -a file:{params.adapter_file} -e 0.1 -O 1 --quiet --minimum-length 18 \" \n",
    "        \"--maximum-length 40 -o {output.cutadapt_reads} {input} > {output.log}\"\n",
    "\n",
    "\n",
    "rule bowtie1_alignment:\n",
    "    input:\n",
    "        reads = 'snakemake_cutadapt/{file}_umi_cutadapt.fastq'\n",
    "    output:\n",
    "        aligned_sam = 'snakemake_bowtie_stringent/aligned/{file}_bowtie_aligned.sam',\n",
    "        non_aligned = 'snakemake_bowtie_stringent/not_aligned/{file}_bowtie_non_aligned.fastq',\n",
    "        log = 'snakemake_bowtie_stringent/reports/{file}_bowtie_aligned.log'\n",
    "    params:\n",
    "        index = 'indexes/bowtie_index/hg_38_index/hg_38'\n",
    "    shell:\n",
    "        \"bowtie -n 0 -l 10 --best --strata -m 1 -k 1 -x {params.index} {input.reads} \"\n",
    "        \"--un {output.non_aligned} -S {output.aligned_sam} 2> {output.log}\" \n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 203,
   "id": "edf772c6-73ea-4641-88fc-63f42d375004",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Overwriting snakefile_bowtie1_miRNA\n"
     ]
    }
   ],
   "source": [
    "%%writefile snakefile_bowtie1_miRNA\n",
    "\n",
    "# Testing the parameters that align to the mirBASE dataset\n",
    "\n",
    "READS = [file.stem for file in Path('SUBSET_DATA/').iterdir() if file.suffix == '.fastq']\n",
    "\n",
    "rule all:\n",
    "    input:\n",
    "        expand('snakemake_bowtie_mirbase/aligned/{file}_bowtie_aligned.sam', file = READS)\n",
    "        \n",
    "    \n",
    "rule remove_umi:\n",
    "    input:\n",
    "        raw_reads = 'SUBSET_DATA/{file}.fastq'\n",
    "    output:\n",
    "        umi_reads = 'snakemake_umi/umi_trimmed/{file}_umi.fastq',\n",
    "        log_file = 'snakemake_umi/reports/{file}_umi.log'\n",
    "    shell:\n",
    "        \"umi_tools extract --extract-method=regex --bc-pattern='(?P<discard_1>.*)(?P<umi_1>[ACT]{{8}}CA)' \" \n",
    "        \"-I {input.raw_reads} -S {output.umi_reads} -L {output.log_file}\" \n",
    "        #FOR REMOVING 3' ADAPTER: '(?P<discard_1>.*)(?P<umi_1>[ACT]{{8}}CA).*(?P<discard_2>TGGAATTCT.+)' \n",
    "        \n",
    "rule cutadapt:\n",
    "    input:\n",
    "        'snakemake_umi/umi_trimmed/{file}_umi.fastq'\n",
    "    output:\n",
    "        cutadapt_reads = 'snakemake_cutadapt/{file}_umi_cutadapt.fastq',\n",
    "        log = 'snakemake_cutadapt/reports/{file}_cutadapt.log'\n",
    "    params:\n",
    "        adapter_file = 'adapters/adapters.txt'\n",
    "    shell:\n",
    "        \"cutadapt -a file:{params.adapter_file} -e 0.1 -O 1 --quiet --minimum-length 18 \" \n",
    "        \"--maximum-length 40 -o {output.cutadapt_reads} {input} > {output.log}\"\n",
    "\n",
    "\n",
    "rule bowtie1_alignment:\n",
    "    input:\n",
    "        reads = 'snakemake_cutadapt/{file}_umi_cutadapt.fastq'\n",
    "    output:\n",
    "        aligned_sam = 'snakemake_bowtie_mirbase/aligned/{file}_bowtie_aligned.sam',\n",
    "        non_aligned = 'snakemake_bowtie_mirbase/not_aligned/{file}_bowtie_non_aligned.fastq',\n",
    "        log = 'snakemake_bowtie_mirbase/reports/{file}_bowtie_aligned.log'\n",
    "    params:\n",
    "        index = 'indexes/bowtie_index/mirbase_all_index/mirbase_all'\n",
    "    shell:\n",
    "        \"bowtie -n 3 -l 10 --best --strata -m 100 -k 1 -x {params.index} {input.reads} \"\n",
    "        \"--un {output.non_aligned} -S {output.aligned_sam} 2> {output.log}\" \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 219,
   "id": "c686fc08-e369-4f24-acbb-4f892b8353fc",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Overwriting snakefile_bowtie2_ziemann\n"
     ]
    }
   ],
   "source": [
    "%%writefile snakefile_bowtie2_ziemann\n",
    "\n",
    "# File that tries bowtie2.vsl from Zieman article in RNA. This is supposed to be the best aligner! \n",
    "# BUT with --end-to-end isntead of --local mode --> this does NOT allow soft clipping!! \n",
    "\n",
    "READS = [file.stem for file in Path('SUBSET_DATA/').iterdir() if file.suffix == '.fastq']\n",
    "\n",
    "rule all:\n",
    "    input:\n",
    "        expand('snakemake_bowtie2/{file}_aligned_bowtie2.bam', file = READS),\n",
    "        'snakemake_featurecounts/featurecounts_genome_clean.csv'\n",
    "        \n",
    "rule remove_umi:\n",
    "    input:\n",
    "        raw_reads = 'SUBSET_DATA/{file}.fastq'\n",
    "    output:\n",
    "        umi_reads = 'snakemake_umi/umi_trimmed/{file}_umi.fastq',\n",
    "        log_file = 'snakemake_umi/reports/{file}_umi.log'\n",
    "    shell:\n",
    "        \"umi_tools extract --extract-method=regex --bc-pattern='(?P<discard_1>.*)(?P<umi_1>[ACT]{{8}}CA)' \" \n",
    "        \"-I {input.raw_reads} -S {output.umi_reads} -L {output.log_file}\" \n",
    "        #FOR REMOVING 3' ADAPTER: '(?P<discard_1>.*)(?P<umi_1>[ACT]{{8}}CA).*(?P<discard_2>TGGAATTCT.+)' \n",
    "        \n",
    "rule cutadapt:\n",
    "    input:\n",
    "        'snakemake_umi/umi_trimmed/{file}_umi.fastq'\n",
    "    output:\n",
    "        cutadapt_reads = 'snakemake_cutadapt/{file}_umi_cutadapt.fastq',\n",
    "        log = 'snakemake_cutadapt/reports/{file}_cutadapt.log'\n",
    "    params:\n",
    "        adapter_file = 'adapters/adapters.txt'\n",
    "    shell:\n",
    "        \"cutadapt -a file:{params.adapter_file} -e 0.1 -O 1 --quiet --minimum-length 18 \" \n",
    "        \"--maximum-length 40 -o {output.cutadapt_reads} {input} > {output.log}\"\n",
    "\n",
    "rule bowtie2_genome_alignment:\n",
    "    input:\n",
    "        reads = 'snakemake_cutadapt/{file}_umi_cutadapt.fastq'\n",
    "    output:\n",
    "        aligned = 'snakemake_bowtie2/{file}_aligned_bowtie2.bam',\n",
    "        log = 'snakemake_bowtie2/reports/{file}.log'\n",
    "    params:\n",
    "        index = 'indexes/bowtie2_index/genome/GRCh38'\n",
    "    shell:\n",
    "        \"bowtie2 -k 100 --end-to-end --very-sensitive -x {params.index} -U {input.reads} 2> {output.log} | \" \n",
    "        \"samtools view -b - | samtools sort -o {output.aligned} - && samtools index {output.aligned}\"\n",
    "        \n",
    "        \n",
    "rule count_reads_genome_featurecounts:\n",
    "    input:\n",
    "        gtf_file = 'mirbase/hsa.gff3'\n",
    "    params:\n",
    "        reads_list = 'snakemake_bowtie2/*bam'\n",
    "    output:\n",
    "        counts = 'snakemake_featurecounts/featurecounts_genome.txt' \n",
    "    shell:\n",
    "        \"featureCounts -t miRNA -g ID -f -O -a {input.gtf_file} -o {output.counts} {params.reads_list}\"\n",
    "        \n",
    "#featureCounts -t miRNA -g ID -f -s 1 -O -T 16 -a /smallRNA.gtf -o /output_dir n.bam 2>featureCounts.log\n",
    "#USE -s 1 or not??????? \n",
    "\n",
    "\n",
    "# clean columns in featurecounts matrix \n",
    "rule clean_featurecounts:\n",
    "    input:\n",
    "        'snakemake_featurecounts/featurecounts_genome.txt'\n",
    "    output:\n",
    "        'snakemake_featurecounts/featurecounts_genome_clean.csv'\n",
    "    script:\n",
    "        'scripts/clean_featurecounts.py'\n",
    "        \n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
